sys/kern/vm_map.c

```diff
@@ -230,18 +230,39 @@ int vm_map_findspace(vm_map_t *map, vaddr_t *start_p, size_t length) {
   return vm_map_findspace_nolock(map, start_p, length, NULL);
 }
 
+static void vm_map_preparespace_nolock(vm_map_t *map, vaddr_t start,
+                                       size_t length) {
+  assert(page_aligned_p(start) && page_aligned_p(length));
+
+  if (!vm_map_contains_p(map, start, start + length))
+    return;
+
+  vm_segment_t *seg, *next;
+  TAILQ_FOREACH_SAFE (seg, &map->entries, link, next) {
+    if ((start <= seg->start && seg->start < start + length) ||
+        (start < seg->end && seg->end < start + length)) {
+      vm_segment_destroy(map, seg);
+    }
+  }
+}
+
 int vm_map_insert(vm_map_t *map, vm_segment_t *seg, vm_flags_t flags) {
   SCOPED_MTX_LOCK(&map->mtx);
   vm_segment_t *after;
   vaddr_t start = seg->start;
   size_t length = seg->end - seg->start;
   vm_seg_flags_t seg_flags = 0;
 
+  if ((flags & VM_FIXED)) {
+    vm_map_preparespace_nolock(map, start, length);
+  }
+
   int error = vm_map_findspace_nolock(map, &start, length, &after);
+
   if (error)
     return error;
-  if ((flags & VM_FIXED) && (start != seg->start))
-    return ENOMEM;
+
+  assert(((flags & VM_FIXED) && (start == seg->start)) || !(flags & VM_FIXED));
```

Comment:
As far as I understand it, the first part of assert will fail when the start argument is not in our map (according to `vm_map_findspace_nolock` implementation cause of `start = max(start, vm_map_start(map));`).
Are you sure about this [comment](https://github.com/cahirwpz/mimiker/pull/907#discussion_r554988238)?

Reply:
Okay, but above the assert is if statement if error is not zero. And if is, then return with it. So, to be honest, I don't know what do you mean exactly.

Reply:
I meant situatuion when `vm_map_start(map) > seg->start` and `VM_FIXED` is set in `vm_map_insert`, does `vm_map_findspace_nolock` always return error in such case?

Reply:
Okay, I see. In that case `vm_map_findspace_nolock` will return some another start address that will be in our `vm_map`. Okay, I agree. It can lead to some problems. I will fix it